File: app/src/main/java/com/example/pokeapitarea/MainActivity.java
package com.example.pokeapitarea;


import android.content.Context;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.util.Log;

import android.view.View;
import android.widget.Toast;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

import java.util.ArrayList;


public class MainActivity extends AppCompatActivity {

    private static final String TAG = "POKEDEX";

    private Retrofit retrofit;


    private RecyclerView recyclerView;
    private PokemonAdapter listaPokemonAdapter;

    private int offset;

    private boolean cargar;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);


        recyclerView = (RecyclerView) findViewById(R.id.rv_pokemon_list);
        listaPokemonAdapter = new PokemonAdapter(this,new ClickListenerRV() {
            @Override
            public void onClick(View v, int position) {
                recyclerView.setBackgroundColor(255);
            }
        });
        recyclerView.setAdapter(listaPokemonAdapter);
        recyclerView.setHasFixedSize(true);
        final LinearLayoutManager layoutManager = new LinearLayoutManager(this);
        recyclerView.setLayoutManager(layoutManager);
        recyclerView.addOnScrollListener(new RecyclerView.OnScrollListener() {
            @Override
            public void onScrolled(RecyclerView recyclerView, int dx, int dy) {
                super.onScrolled(recyclerView, dx, dy);

                if (dy > 0) {
                    int visibleItemCount = layoutManager.getChildCount();
                    int totalItemCount = layoutManager.getItemCount();
                    int pastVisibleItems = layoutManager.findFirstVisibleItemPosition();

                    if (cargar) {
                        if ((visibleItemCount + pastVisibleItems) >= totalItemCount) {
                            Log.i(TAG, " Llegamos al final.");

                            cargar = false;
                            offset += 20;
                            getData(offset);
                        }
                    }
                }
            }
        });


        retrofit = new Retrofit.Builder()
                .baseUrl("http://pokeapi.co/api/v2/")
                .addConverterFactory(GsonConverterFactory.create())
                .build();

        cargar = true;
        offset = 0;
        getData(offset);
    }

    private void getData(int offset) {
        PokeNet service = retrofit.create(PokeNet.class);
        Call<Respond> pokemonRespuestaCall = service.obtenerListaPokemon(20, offset);

        pokemonRespuestaCall.enqueue(new Callback<Respond>() {
            @Override
            public void onResponse(Call<Respond> call, Response<Respond> response) {
                cargar = true;
                if (response.isSuccessful()) {

                    Respond pokemonRespuesta = response.body();
                    ArrayList<Pokemon> listaPokemon = pokemonRespuesta.getResults();

                    listaPokemonAdapter.addPokemon(listaPokemon);

                } else {
                    Log.e(TAG, " onResponse: " + response.errorBody());
                }
            }

            @Override
            public void onFailure(Call<Respond> call, Throwable t) {
                cargar = true;
                Log.e(TAG, " onFailure: " + t.getMessage());
            }
        });
    }
}

File: app/src/main/java/com/example/pokeapitarea/MainActivity.kt
package com.example.pokeapitarea


import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.util.Log
import android.view.View
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory


class MainActivity : AppCompatActivity() {

    private var retrofit: Retrofit? = null


    private var recyclerView: RecyclerView? = null
    private var listaPokemonAdapter: PokemonAdapter? = null

    private var offset: Int = 0

    private var cargar: Boolean = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)


        recyclerView = findViewById<View>(R.id.rv_pokemon_list) as RecyclerView
        listaPokemonAdapter = PokemonAdapter(this)

        recyclerView!!.adapter = listaPokemonAdapter
        recyclerView!!.setHasFixedSize(true)
        val layoutManager = LinearLayoutManager(this)
        recyclerView!!.layoutManager = layoutManager
        recyclerView!!.addOnScrollListener(object : RecyclerView.OnScrollListener() {
            override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
                super.onScrolled(recyclerView, dx, dy)

                if (dy > 0) {
                    val visibleItemCount = layoutManager.childCount
                    val totalItemCount = layoutManager.itemCount
                    val pastVisibleItems = layoutManager.findFirstVisibleItemPosition()

                    if (cargar) {
                        if (visibleItemCount + pastVisibleItems >= totalItemCount) {
                            Log.i(TAG, " Llegamos al final.")

                            cargar = false
                            offset += 20
                            getData(offset)
                        }
                    }
                }
            }
        })


        retrofit = Retrofit.Builder()
            .baseUrl("http://pokeapi.co/api/v2/")
            .addConverterFactory(GsonConverterFactory.create())
            .build()

        cargar = true
        offset = 0
        getData(offset)
    }

    private fun getData(offset: Int) {
        val service = retrofit!!.create(PokeNet::class.java)
        val pokemonRespuestaCall = service.obtenerListaPokemon(20, offset)

        pokemonRespuestaCall!!.enqueue(object : Callback<Respond> {
            override fun onResponse(call: Call<Respond>, response: Response<Respond>) {
                cargar = true
                if (response.isSuccessful) {

                    val pokemonRespuesta = response.body()
                    val listaPokemon = pokemonRespuesta!!.results

                    listaPokemonAdapter!!.addPokemon(listaPokemon!!)

                } else {
                    Log.e(TAG, " onResponse: " + response.errorBody()!!)
                }
            }

            override fun onFailure(call: Call<Respond>, t: Throwable) {
                cargar = true
                Log.e(TAG, " onFailure: " + t.message)
            }
        })
    }

    companion object {

        private val TAG = "POKEDEX"
    }
}